packages/nodes-base/nodes/Code/Code.node.ts

```diff
@@ -0,0 +1,140 @@
+import type { CodeNodeMode } from './utils';
+
+export class Code implements INodeType {
+	description: INodeTypeDescription = {
+		displayName: 'Code',
+		name: 'code',
+		icon: 'fa:code',
+		group: ['transform'],
+		version: 1,
+		description: 'Run custom JavaScript code',
+		defaults: {
+			name: 'Code',
+			color: '#FF9922',
+		},
+		inputs: ['main'],
+		outputs: ['main'],
+		parameterPane: 'wide',
+		properties: [
+			{
+				displayName: 'Mode',
+				name: 'mode',
+				type: 'options',
+				noDataExpression: true,
+				options: [
+					{
+						name: 'Run Once for All Items',
+						value: 'runOnceForAllItems',
+						description: 'Run this code only once, no matter how many input items there are',
+					},
+					{
+						name: 'Run Once for Each Item',
+						value: 'runOnceForEachItem',
+						description: 'Run this code as many times as there are input items',
+					},
+				],
+				default: 'runOnceForAllItems',
+			},
+			{
+				displayName: 'JavaScript',
+				name: 'jsCode',
+				typeOptions: {
+					editor: 'codeNodeEditor',
+				},
+				type: 'string',
+				default: '', // set by component
+				description:
+					'JavaScript code to execute.<br><br>Tip: You can use luxon vars like <code>$today</code> for dates and <code>$jmespath</code> for querying JSON structures. <a href="https://docs.n8n.io/nodes/n8n-nodes-base.function">Learn more</a>.',
+				noDataExpression: true,
+			},
+			{
+				displayName:
+					'Type <code>$</code> for a list of special vars/functions. Debug using <code>console.log()</code> statements and viewing their output in the browser console. <a>More info</a>',
+				name: 'notice',
+				type: 'notice',
+				default: '',
+			},
+		],
+	};
+
+	async execute(this: IExecuteFunctions) {
+		let items = this.getInputData();
+
+		const nodeMode = this.getNodeParameter('mode', 0) as CodeNodeMode;
+		const workflowMode = this.getMode();
+
+		// ----------------------------------
+		//        runOnceForAllItems
+		// ----------------------------------
+
+		if (nodeMode === 'runOnceForAllItems') {
+			items = deepCopy(items);
+
+			const jsCode = this.getNodeParameter('jsCode', 0) as string;
+
+			const context = getSandboxContext.call(this);
+			const sandbox = new Sandbox(context, workflowMode, nodeMode);
+
+			if (workflowMode === 'manual') {
+				sandbox.on('console.log', this.sendMessageToUI);
+			}
+
+			try {
+				items = await sandbox.runCode(jsCode);
+			} catch (error) {
+				if (!this.continueOnFail()) return Promise.reject(error);
+				items = [{ json: { error: error.message } }];
+			}
+
+			for (const item of items) {
+				standardizeOutput(item.json);
+			}
+
+			return this.prepareOutputData(items);
+		}
+
+		// ----------------------------------
+		//        runOnceForEachItem
+		// ----------------------------------
+
+		const returnData: INodeExecutionData[] = [];
+
+		for (let index = 0; index < items.length; index++) {
+			let item = deepCopy(items[index]);
```

Comment:
Did the copy stay behind from old code? Because right now it looks like we are making a very expensive copy of the data and then actually never using it.

Reply:
Addressed in [07a3819](https://github.com/n8n-io/n8n/pull/3965/commits/07a3819ae9140875b96344b16cbea1002cbcb40f)